plugins/pulp_rpm/plugins/importers/yum/purge.py

```diff
@@ -400,55 +400,40 @@ def _duplicate_key_id_generator_aggregation(unit, fields):
     # It is a happy coincidence that the two dicts are the same values
     pipeline_opts = dict(zip(fields, repeat(1)))
 
-    # for $sort, this indicates ascending sort for nevra fields
-    sort = {'$sort': pipeline_opts}
+    lookup = {'$lookup': {'from': unit._meta['collection'],
+                          'localField': 'unit_id', 'foreignField': '_id', 'as': 'units'}}
+
+    match = {'$match': {'repo_id': repo_id, 'unit_type_id': unit._content_type_id.default}}
+
+    unwind = {'$unwind': '$units'}
+
+    replace_root = {'$replaceRoot': {'newRoot': '$units'}}
 
     # for $project, this indicates what fields to include in the result
     project = {'$project': pipeline_opts}
 
     # When aggregating over hundreds of thousands of packages, mongo can overflow
     # To prevent this, mongo needs to be allowed to temporarily use the disk for this transaction
-    # Set the batch size to 5 to prevent a cursor timeout
-    aggregation = unit.objects.aggregate(sort, project, allowDiskUse=True,
-                                         batchSize=5)
-
-    # loop state tracking vars
-    previous_nevra = None
-    previous_pkg_id = None
-    yielding_ids = None
-
+    # It should be ok to set the batch size to 100 as it is not doing sorting. The results should
+    # be returned within 100ms to a few seconds (if the IO is really slow).
+    # Simply reduce the batchSize if you are getting the cursor timeout. According to my tests,
+    # the getMore query is performed slower when setting low batchSize like 5.
+    aggregation = model.RepositoryContentUnit.objects.aggregate(lookup, match, unwind,
+                                         replace_root, project, allowDiskUse=True,
+                                         batchSize=100)
+
+    package_group = {}
```

Comment:
I'd prefer we use ```collections.defaultdict(list)``` for better clarity here.

Reply:
Done. Thanks for the suggestion